src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -14,592 +14,549 @@ namespace Microsoft.CodeAnalysis.CSharp
+                var concatMethods = new WellKnownConcatRelatedMethods(_compilation);
+                AddConcatArgument(unvisitedRight, argumentAlreadyVisited: false, arguments, ref concatMethods);
+                AddConcatArgument(left, argumentAlreadyVisited: true, arguments, ref concatMethods);
+                arguments.ReverseContents();
             }
 
-            // Convert both sides to a string (calling ToString if necessary)
-            loweredLeft = ConvertConcatExprToString(loweredLeft);
-            loweredRight = ConvertConcatExprToString(loweredRight);
+            return RewriteStringConcat(syntax, arguments);
+        }
 
-            Debug.Assert(loweredLeft.Type is { } && (loweredLeft.Type.IsStringType() || loweredLeft.Type.IsErrorType()) || loweredLeft.ConstantValueOpt?.IsNull == true);
-            Debug.Assert(loweredRight.Type is { } && (loweredRight.Type.IsStringType() || loweredRight.Type.IsErrorType()) || loweredRight.ConstantValueOpt?.IsNull == true);
+        private BoundExpression VisitStringConcatenation(BoundBinaryOperator originalOperator)
+        {
+            Debug.Assert(IsBinaryStringConcatenation(originalOperator));
 
-            // try fold two args without flattening.
-            var folded = TryFoldTwoConcatOperands(loweredLeft, loweredRight);
-            if (folded != null)
+            if (_inExpressionLambda)
             {
-                return folded;
+                // Case 1, expression trees
+                return RewriteStringConcatInExpressionLambda(
+                    originalOperator,
+                    compoundAssignmentLeft: null,
+                    compoundAssignmentOperatorKind: null,
+                    compoundAssignmentSyntax: null,
+                    compoundAssignmentType: null);
             }
 
-            // flatten and merge -  ( expr1 + "A" ) + ("B" + expr2) ===> (expr1 + "AB" + expr2)
-            ArrayBuilder<BoundExpression> leftFlattened = ArrayBuilder<BoundExpression>.GetInstance();
-            ArrayBuilder<BoundExpression> rightFlattened = ArrayBuilder<BoundExpression>.GetInstance();
+            // We'll walk the children in a depth-first order, pull all the arguments out, and then visit them. We'll fold any constant arguments as
+            // we go, pulling them all into a string literal.
+            CollectAndVisitConcatArguments(originalOperator, visitedCompoundAssignmentLeftRead: null, out var arguments);
 
-            FlattenConcatArg(loweredLeft, leftFlattened);
-            FlattenConcatArg(loweredRight, rightFlattened);
+            return RewriteStringConcat(originalOperator.Syntax, arguments);
+        }
 
-            if (leftFlattened.Any() && rightFlattened.Any())
+        private BoundExpression RewriteStringConcat(SyntaxNode originalSyntax, ArrayBuilder<BoundExpression> visitedArguments)
+        {
+            // The strategy of this rewrite is to pull out all the individual arguments of the concat, visit them as appropriate,
```

Comment:
>to pull out all the individual arguments of the concat, visit them as appropriate, [](http://example.com/codeflow?start=47&length=82)

It doesn't look like this method is responsible for doing so. Perhaps we should say that we expect the caller to prepare the input this way?